### DISCUSSION TITLE
Multiple Dispatch can Implement Object Oriented, but not visa versa

### DISCUSSION THREAD

**Discussion by dehann** [JuliaRobotics/Caesar.jl#779] [General]
This is to show Julia can implement this pattern, but better patterns exist.  Don't use this as a design guide, more as info for public benefit.

```julia
function sumUtil(state, args...)
  @info "sumUtil" length(state) args
  sum(state)
end

struct MyClass
  # public object state
  state::Vector{Float64}

  # member functions
  sumValues::Function

  MyClass(;state=randn(10)) = new(state, (args...)->sumUtil(state, args...))
end



obj = MyClass();

obj.sumValues("Hello world") 
```

Also see Stefan's K. talk:
https://www.youtube.com/watch?v=kc9HwsxE1OY

**Comment by dehann** [comment]
Julia Modules are very similar to C++ namespaces, and struct  is the main mechanism to do encapsulation.  The difference (allowing for multiple dispatch) being that Julia allows for functional programming, 'all-three-closures', and dynamic-to-static type inference,

**Reply by dehann** [reply]
'all-three-closures' (using lexical scope) implies that globals/outer/namespace functions, inner functions, and lambda functions can access any state within lexical scope:
https://docs.julialang.org/en/v1/manual/variables-and-scoping/#scope-of-variables

this should* all compile down automatically to a type-stable static functions (one static function each for different argument dispatches, so many potential functions from this one method definition)

```julia
function QuickScope(userdata)
  inner(x,y) = x + y + userdata
  objective = (x) -> inner(x, 3)
  
  val = randn()
  objective(val) - objective(val + 0.001)
end
```

**Comment by GearsAD** [comment]
Is this class pattern used by the Julia developers? I don't think I've seen it around, concerned that we would use it for the SDK standardization and it'll be strange for Julia users.

**Reply by dehann** [reply]
Oh, this pattern would be considered bad Julia, just showing it's possible

**Comment by GearsAD** [comment]
If the intention is to boil down what we should consider as supported patterns in Julia, I'm not convinced that classical OOP is supported (it is, but by a bad Julia pattern). 

I'd be more comfortable saying "we don't do this here".
